File: internal/daemon/scheduler/scheduler.go
// Package scheduler is the daemon's job queue + worker pool.
//
// Per plan §6 / docs/plans/20260518-Daemon-UDS-Plan.md §4.2: jobs are
// enqueued as qualified pairs (project, job_id), picked up by a fixed-
// size worker pool, and handed to an Executor that drives the per-project
// runner and transitions the run through MarkRunning → terminal.
//
// The scheduler itself no longer owns the restart-recovery sweep — that
// is now run per-project by internal/daemon/projectmgr when a project is
// opened for the first time. The scheduler also does not own a runstore
// handle (there is no globally-shared one anymore); the queue is project-
// agnostic and only carries opaque qualified ids.
package scheduler

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"sync"
)

// Job is a qualified job reference. Project is the canonical absolute
// project root (matches projectmgr.Key) and ID is the per-project
// daemon_runs.job_id.
type Job struct {
	Project string
	ID      string
}

// String renders a Job as "<project>::<id>" — used as the internal map
// key and for logging.
func (j Job) String() string { return j.Project + "::" + j.ID }

// Executor is the per-job lifecycle implementation. The daemon wires
// this as a closure that looks up the project in the manager and calls
// proj.Executor.Run(ctx, job.ID). Tests substitute a mock.
//
// Implementations are responsible for the full state machine: MarkRunning,
// MarkDone/MarkFailed/MarkCancelled. The scheduler observes the return
// value only for logging — terminal state must already be persisted.
type Executor interface {
	Run(ctx context.Context, job Job) error
}

// ExecutorFunc adapts a plain function to the Executor interface.
type ExecutorFunc func(ctx context.Context, job Job) error

// Run implements Executor.
func (f ExecutorFunc) Run(ctx context.Context, job Job) error { return f(ctx, job) }

// Config tunes the scheduler.
type Config struct {
	// Workers is the maximum number of concurrent jobs. <= 0 → 1.
	Workers int
	// QueueDepth is the buffered channel size. <= 0 → max(Workers, 16).
	QueueDepth int
	// Logger receives info/warn output. nil → slog.Default().
	Logger *slog.Logger
}

// Scheduler coordinates a worker pool against an Executor.
type Scheduler struct {
	exec Executor
	cfg  Config
	log  *slog.Logger

	queue chan Job

	startMu sync.Mutex
	started bool
	stopped chan struct{}

	cancelCtx context.CancelFunc
	rootCtx   context.Context

	workersWG sync.WaitGroup

	activeMu sync.Mutex
	active   map[string]context.CancelFunc // keyed by Job.String()
}

// Sentinel errors.
var (
	ErrNotStarted     = errors.New("scheduler: not started")
	ErrAlreadyStarted = errors.New("scheduler: already started")
	ErrJobNotActive   = errors.New("scheduler: job not active")
	ErrQueueFull      = errors.New("scheduler: queue full")
)

// New constructs a scheduler. The caller still has to call Start.
//
// The runstore is intentionally absent: restart recovery is owned by the
// project manager which opens its per-project runstore exactly once.
func New(exec Executor, cfg Config) *Scheduler {
	if cfg.Workers <= 0 {
		cfg.Workers = 1
	}
	if cfg.QueueDepth <= 0 {
		if cfg.Workers > 16 {
			cfg.QueueDepth = cfg.Workers
		} else {
			cfg.QueueDepth = 16
		}
	}
	log := cfg.Logger
	if log == nil {
		log = slog.Default()
	}
	return &Scheduler{
		exec:    exec,
		cfg:     cfg,
		log:     log,
		queue:   make(chan Job, cfg.QueueDepth),
		stopped: make(chan struct{}),
		active:  make(map[string]context.CancelFunc),
	}
}

// Start launches the worker pool. Returns ErrAlreadyStarted on the
// second call.
func (s *Scheduler) Start(ctx context.Context) error {
	s.startMu.Lock()
	defer s.startMu.Unlock()
	if s.started {
		return ErrAlreadyStarted
	}
	s.rootCtx, s.cancelCtx = context.WithCancel(context.Background())
	for i := 0; i < s.cfg.Workers; i++ {
		s.workersWG.Add(1)
		go s.workerLoop(i)
	}
	s.started = true
	return nil
}

// Stop drains the queue and waits for in-flight jobs to settle.
//
//  1. Closes the queue so no new jobs are accepted.
//  2. Cancels every active job context (so executors can observe and
//     unwind: Abort → SIGTERM → MarkCancelled).
//  3. Waits for workers to exit, or for graceCtx to expire.
func (s *Scheduler) Stop(graceCtx context.Context) error {
	s.startMu.Lock()
	if !s.started {
		s.startMu.Unlock()
		return ErrNotStarted
	}
	s.started = false
	s.cancelCtx()
	close(s.queue) // workers exit when queue drains
	s.startMu.Unlock()

	// Cancel every active job.
	s.activeMu.Lock()
	for _, cancel := range s.active {
		cancel()
	}
	s.activeMu.Unlock()

	done := make(chan struct{})
	go func() {
		s.workersWG.Wait()
		close(done)
	}()
	select {
	case <-done:
		close(s.stopped)
		return nil
	case <-graceCtx.Done():
		close(s.stopped)
		return graceCtx.Err()
	}
}

// Stopped returns a channel that is closed once Stop completes.
func (s *Scheduler) Stopped() <-chan struct{} { return s.stopped }

// Enqueue places a qualified job in the queue. Returns ErrQueueFull if
// the buffered channel is at capacity (caller is responsible for back
// pressure handling at the HTTP/poller layer).
//
// startMu is held for the duration of the non-blocking send so Stop
// (which closes s.queue while holding the same lock) cannot race the
// send into a closed channel. The send itself is non-blocking, so the
// critical section is bounded by a single channel op + map lookup.
func (s *Scheduler) Enqueue(job Job) error {
	s.startMu.Lock()
	defer s.startMu.Unlock()
	if !s.started {
		return ErrNotStarted
	}
	if job.ID == "" {
		return fmt.Errorf("scheduler: empty job id")
	}
	select {
	case s.queue <- job:
		return nil
	default:
		return ErrQueueFull
	}
}

// Cancel signals the per-job context for the qualified job. The executor
// is expected to observe and unwind. Returns ErrJobNotActive if no such
// job is in flight (the caller can still PATCH the row directly if
// needed; the scheduler does not touch the DB on cancel).
func (s *Scheduler) Cancel(job Job) error {
	s.activeMu.Lock()
	cancel, ok := s.active[job.String()]
	s.activeMu.Unlock()
	if !ok {
		return ErrJobNotActive
	}
	cancel()
	return nil
}

// IsActive reports whether the named job is currently being executed.
func (s *Scheduler) IsActive(job Job) bool {
	s.activeMu.Lock()
	defer s.activeMu.Unlock()
	_, ok := s.active[job.String()]
	return ok
}

// ActiveJobs returns a snapshot of the currently active jobs.
func (s *Scheduler) ActiveJobs() []Job {
	s.activeMu.Lock()
	defer s.activeMu.Unlock()
	out := make([]Job, 0, len(s.active))
	for k := range s.active {
		// Decode "<project>::<id>" back into a Job. We do not bother
		// validating — the keys are only ever inserted by runOne.
		if idx := indexOfDoubleColon(k); idx >= 0 {
			out = append(out, Job{Project: k[:idx], ID: k[idx+2:]})
		}
	}
	return out
}

// indexOfDoubleColon finds the first "::" in s, or -1.
func indexOfDoubleColon(s string) int {
	for i := 0; i+1 < len(s); i++ {
		if s[i] == ':' && s[i+1] == ':' {
			return i
		}
	}
	return -1
}

// ActiveCountByProject returns the number of in-flight jobs grouped by
// project key (used for the aggregated health endpoint).
func (s *Scheduler) ActiveCountByProject() map[string]int {
	s.activeMu.Lock()
	defer s.activeMu.Unlock()
	out := make(map[string]int, len(s.active))
	for k := range s.active {
		if idx := indexOfDoubleColon(k); idx >= 0 {
			out[k[:idx]]++
		}
	}
	return out
}

func (s *Scheduler) workerLoop(idx int) {
	defer s.workersWG.Done()
	for {
		select {
		case <-s.rootCtx.Done():
			return
		case job, ok := <-s.queue:
			if !ok {
				return
			}
			s.runOne(idx, job)
		}
	}
}

func (s *Scheduler) runOne(workerIdx int, job Job) {
	jobCtx, cancel := context.WithCancel(s.rootCtx)
	key := job.String()
	s.activeMu.Lock()
	s.active[key] = cancel
	s.activeMu.Unlock()
	defer func() {
		s.activeMu.Lock()
		delete(s.active, key)
		s.activeMu.Unlock()
		cancel()
	}()

	s.log.Info("scheduler: run start", "worker", workerIdx, "job", key)
	err := s.exec.Run(jobCtx, job)
	if err != nil {
		s.log.Warn("scheduler: executor returned error", "job", key, "err", err)
	} else {
		s.log.Info("scheduler: run finished", "job", key)
	}
}

File: crates/autosk-core/src/scheduler.rs
//! Global job queue + worker pool — the Rust port of
//! `internal/daemon/scheduler`.
//!
//! Jobs are enqueued as qualified `(project, job_id)` pairs, picked up by a
//! fixed-size worker pool, and handed to a [`SchedExecutor`] that drives the
//! run through its terminal state. Cancellation fires the per-job
//! [`CancelToken`]; the executor observes it and unwinds.

use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::mpsc::{sync_channel, Receiver, SyncSender, TrySendError};
use std::sync::{Arc, Mutex};
use std::thread::JoinHandle;

use crate::ctx::{CancelToken, Ctx};

/// A qualified job reference (mirror of `scheduler.Job`).
#[derive(Debug, Clone)]
pub struct Job {
    /// Canonical project root (matches the project manager's key).
    pub project: String,
    /// Per-project `daemon_runs.job_id`.
    pub id: String,
}

impl Job {
    /// `"<project>::<id>"` — the internal active-map key.
    pub fn key(&self) -> String {
        format!("{}::{}", self.project, self.id)
    }
}

/// The per-job lifecycle implementation the scheduler drives. The daemon wires
/// this to look up the job's project and call its executor (mirror of the Go
/// `scheduler.Executor` interface).
pub trait SchedExecutor: Send + Sync {
    /// Runs the job to a terminal state. The scheduler only logs the return.
    fn run(&self, ctx: &Ctx, job: &Job);
}

/// Scheduler tuning (mirror of `scheduler.Config`).
#[derive(Debug, Clone)]
pub struct Config {
    pub workers: usize,
    pub queue_depth: usize,
}

impl Default for Config {
    fn default() -> Self {
        Config {
            workers: 1,
            queue_depth: 16,
        }
    }
}

/// Why an enqueue failed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EnqueueError {
    NotStarted,
    QueueFull,
    EmptyId,
}

/// The scheduler.
pub struct Scheduler {
    exec: Arc<dyn SchedExecutor>,
    tx: SyncSender<Job>,
    rx: Mutex<Option<Receiver<Job>>>,
    root: Ctx,
    cancel_root: CancelToken,
    started: AtomicBool,
    workers: Mutex<Vec<JoinHandle<()>>>,
    active: Arc<Mutex<HashMap<String, CancelToken>>>,
    num_workers: usize,
}

impl Scheduler {
    /// Constructs a scheduler. Call [`Scheduler::start`] to launch the pool.
    pub fn new(exec: Arc<dyn SchedExecutor>, cfg: Config) -> Arc<Scheduler> {
        let workers = cfg.workers.max(1);
        let depth = if cfg.queue_depth == 0 {
            workers.max(16)
        } else {
            cfg.queue_depth
        };
        let (tx, rx) = sync_channel(depth);
        let (root, cancel_root) = Ctx::new_cancellable();
        Arc::new(Scheduler {
            exec,
            tx,
            rx: Mutex::new(Some(rx)),
            root,
            cancel_root,
            started: AtomicBool::new(false),
            workers: Mutex::new(Vec::new()),
            active: Arc::new(Mutex::new(HashMap::new())),
            num_workers: workers,
        })
    }

    /// Launches the worker pool.
    pub fn start(self: &Arc<Self>) {
        if self.started.swap(true, Ordering::SeqCst) {
            return;
        }
        let rx = self.rx.lock().unwrap().take().expect("rx taken once");
        let rx = Arc::new(Mutex::new(rx));
        let mut handles = Vec::new();
        for i in 0..self.num_workers {
            let me = Arc::clone(self);
            let rx = Arc::clone(&rx);
            handles.push(std::thread::spawn(move || me.worker_loop(i, rx)));
        }
        *self.workers.lock().unwrap() = handles;
    }

    /// Enqueues a qualified job (non-blocking; [`EnqueueError::QueueFull`] when
    /// the buffer is full — the caller leaves the row queued for the next tick).
    pub fn enqueue(&self, job: Job) -> Result<(), EnqueueError> {
        if !self.started.load(Ordering::SeqCst) {
            return Err(EnqueueError::NotStarted);
        }
        if job.id.is_empty() {
            return Err(EnqueueError::EmptyId);
        }
        match self.tx.try_send(job) {
            Ok(()) => Ok(()),
            Err(TrySendError::Full(_)) => Err(EnqueueError::QueueFull),
            Err(TrySendError::Disconnected(_)) => Err(EnqueueError::NotStarted),
        }
    }

    /// Fires the cancel token for an active job. No-op when not active.
    pub fn cancel(&self, job: &Job) -> bool {
        if let Some(tok) = self.active.lock().unwrap().get(&job.key()) {
            tok.cancel();
            true
        } else {
            false
        }
    }

    /// Whether the named job is currently executing.
    pub fn is_active(&self, job: &Job) -> bool {
        self.active.lock().unwrap().contains_key(&job.key())
    }

    /// In-flight job count grouped by project (for `healthz`).
    pub fn active_count_by_project(&self) -> HashMap<String, i64> {
        let mut out = HashMap::new();
        for k in self.active.lock().unwrap().keys() {
            if let Some(idx) = k.find("::") {
                *out.entry(k[..idx].to_string()).or_insert(0) += 1;
            }
        }
        out
    }

    /// Cancels every in-flight job and stops accepting new ones. Workers drain
    /// the queue and exit; this joins them.
    pub fn stop(&self) {
        if !self.started.swap(false, Ordering::SeqCst) {
            return;
        }
        self.cancel_root.cancel();
        for tok in self.active.lock().unwrap().values() {
            tok.cancel();
        }
        // Dropping the only sender would close the queue; but we hold `tx`
        // for the scheduler's lifetime, so instead workers exit on the root
        // cancel observed between jobs.
        let handles: Vec<_> = std::mem::take(&mut *self.workers.lock().unwrap());
        for h in handles {
            let _ = h.join();
        }
    }

    fn worker_loop(self: Arc<Self>, _idx: usize, rx: Arc<Mutex<Receiver<Job>>>) {
        loop {
            if self.root.is_cancelled() {
                return;
            }
            let job = {
                let guard = rx.lock().unwrap();
                guard.recv_timeout(std::time::Duration::from_millis(50))
            };
            match job {
                Ok(job) => self.run_one(job),
                Err(std::sync::mpsc::RecvTimeoutError::Timeout) => continue,
                Err(std::sync::mpsc::RecvTimeoutError::Disconnected) => return,
            }
        }
    }

    fn run_one(&self, job: Job) {
        let key = job.key();
        // Each job gets its own cancellable ctx; `cancel(job)` and `stop()`
        // (which fires every active token) both unwind a running executor.
        let (job_ctx, token) = Ctx::new_cancellable();
        self.active.lock().unwrap().insert(key.clone(), token);
        self.exec.run(&job_ctx, &job);
        self.active.lock().unwrap().remove(&key);
    }
}
